core/trino-main/src/main/java/io/trino/execution/scheduler/BinPackingNodeAllocatorService.java

```diff
@@ -361,14 +364,19 @@ public ReserveResult tryReserve(PendingAcquire acquire)
                         }
                         return false;
                     })
-                    .max(comparing(node -> nodesRemainingMemory.get(node.getNodeIdentifier())));
+                    .max(comparing(node -> nodesUnreservedMemory.get(node.getNodeIdentifier())));
 
             if (selectedNode.isEmpty()) {
                 return ReserveResult.NONE_MATCHING;
             }
 
             boolean selectedCandidateMatches = false;
-            if (nodesRemainingMemory.get(selectedNode.get().getNodeIdentifier()) >= acquire.getMemoryLease() || isNodeEmpty(selectedNode.get())) {
+
+            String selectNodeId = selectedNode.get().getNodeIdentifier();
+            boolean enoughUnreservedMemory = nodesUnreservedMemory.get(selectNodeId) >= acquire.getMemoryLease();
+            boolean enoughRuntimeMemory = !nodeMemoryPoolInfos.containsKey(selectNodeId)
+                    || nodeMemoryPoolInfos.get(selectNodeId).getFreeBytes() >= acquire.getMemoryLease();
```

Comment:
I wonder if it makes sense to leave some "head room", e.g. some percentage of the memory pool size? (say 10-15%)

Reply:
Maybe. I would not introduce extra parameter yet though. The more we have of those the less obvious it is what is the correct combination. 

Actually, the bigger question here is should we account assigned tasks (not yet running) towards "used" memory. Currently we do not.
So if we have a node N with:
* 10 GB of mem total
* 4 GB reserved for one task already running
* 8 GB  actually used by task already running.

I will still allow reserving extra 3 tasks each requesting 2GB of memory. Maybe I should only allow for 1? WDYT (let's chat about that today).

Reply:
Does it matter? Anyways at the end only one task will be granted. So doesn't feel like a problem

Reply:
Not sure if I understand the question. But maybe it does not matter as impl is different now :P PTAL